x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/dataframe/transforms/DataFrameTransformCheckpointingInfo.java

```diff
@@ -0,0 +1,137 @@
+            new ConstructingObjectParser<>(
+                    "data_frame_transform_checkpointing_info", true, a -> {
+                        long behind = a[2] == null ? 0L : (Long) a[2];
+
+                        return new DataFrameTransformCheckpointingInfo(
+                                a[0] == null ? DataFrameTransformCheckpointStats.EMPTY : (DataFrameTransformCheckpointStats) a[0],
+                                a[1] == null ? DataFrameTransformCheckpointStats.EMPTY : (DataFrameTransformCheckpointStats) a[1], behind);
+                    });
+
+    static {
+        LENIENT_PARSER.declareObject(ConstructingObjectParser.optionalConstructorArg(),
+                (p, c) -> DataFrameTransformCheckpointStats.fromXContent(p), CURRENT_CHECKPOINT);
+        LENIENT_PARSER.declareObject(ConstructingObjectParser.optionalConstructorArg(),
+                (p, c) -> DataFrameTransformCheckpointStats.fromXContent(p), IN_PROGRESS_CHECKPOINT);
+        LENIENT_PARSER.declareLong(ConstructingObjectParser.optionalConstructorArg(), OPERATIONS_BEHIND);
+    }
+
+    /**
+     * Create checkpoint stats object with checkpoint information about the current and in progress checkpoint as well as the current state
+     * of source.
+     *
+     * @param current stats of the current checkpoint
+     * @param inProgress stats of the in progress checkpoint
+     * @param operationsBehind counter of operations the current checkpoint is behind source
+     */
+    public DataFrameTransformCheckpointingInfo(DataFrameTransformCheckpointStats current, DataFrameTransformCheckpointStats inProgress,
+            long operationsBehind) {
+        this.current = Objects.requireNonNull(current);
+        this.inProgress = Objects.requireNonNull(inProgress);
+        this.operationsBehind = operationsBehind;
+    }
+
+    public DataFrameTransformCheckpointingInfo(StreamInput in) throws IOException {
+        current = new DataFrameTransformCheckpointStats(in);
+        inProgress = new DataFrameTransformCheckpointStats(in);
+        operationsBehind = in.readLong();
+    }
+
+    public DataFrameTransformCheckpointStats getCurrent() {
+        return current;
+    }
+
+    public DataFrameTransformCheckpointStats getInProgress() {
+        return inProgress;
+    }
+
+    public long getOperationsBehind() {
+        return operationsBehind;
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        if (current.getTimestampMillis() > 0) {
+            builder.field(CURRENT_CHECKPOINT.getPreferredName(), current);
+        }
+        if (inProgress.getTimestampMillis() > 0) {
+            builder.field(IN_PROGRESS_CHECKPOINT.getPreferredName(), inProgress);
+        }
+        if (operationsBehind > 0) {
+            builder.field(OPERATIONS_BEHIND.getPreferredName(), operationsBehind);
```

Comment:
it might be good to always include `operations_behind` if there is a current checkpoint. I don't see the harm in `operations_behind: 0`, unless I am misunderstanding the API design.